pkg/autoscaler/autoscaler.go

```diff
@@ -254,295 +256,406 @@ var _ = Describe("[Feature:Machines] Autoscaler should", func() {
+				cleanupObjects[machineSets[i].Name] = machineSets[i]
 			}
-			return len(allNewNodes) == len(machineSets) && framework.NodesAreReady(allNewNodes)
-		}, framework.WaitLong, pollingInterval).Should(BeTrue())
 
-		// Now that we have created some transient machinesets
-		// take stock of the number of nodes we now have.
-		By("Getting nodes")
-		nodes, err := framework.GetNodes(client)
-		Expect(err).NotTo(HaveOccurred())
-		Expect(len(nodes)).To(BeNumerically(">=", 1))
-
-		var machineAutoscalers []*caov1beta1.MachineAutoscaler
+			By(fmt.Sprintf("Creating %v transient machinesets", len(machineSets)))
+			testDuration := time.Now().Add(time.Duration(framework.WaitLong))
+			Eventually(func() bool {
+				By(fmt.Sprintf("[%s remaining] Waiting for nodes to be Ready in %v transient machinesets",
+					remaining(testDuration), len(machineSets)))
+				var allNewNodes []*corev1.Node
+				for i := 0; i < len(machineSets); i++ {
+					nodes, err := framework.GetNodesFromMachineSet(client, machineSets[i])
+					if err != nil {
+						return false
+					}
+					allNewNodes = append(allNewNodes, nodes...)
+				}
+				return len(allNewNodes) == len(machineSets) && framework.NodesAreReady(allNewNodes)
+			}, framework.WaitLong, pollingInterval).Should(BeTrue())
 
-		By(fmt.Sprintf("Creating %v machineautoscalers", len(machineSets)))
-		var clusterExpansionSize int
-		for i := range machineSets {
-			clusterExpansionSize++
-			glog.Infof("Create MachineAutoscaler backed by MachineSet %s/%s - min:%v, max:%v", machineSets[i].Namespace, machineSets[i].Name, 1, 2)
-			asr := machineAutoscalerResource(machineSets[i], 1, 2)
-			Expect(client.Create(ctx, asr)).Should(Succeed())
-			machineAutoscalers = append(machineAutoscalers, asr)
-			cleanupObjects[asr.Name] = asr
-		}
-		Expect(clusterExpansionSize).To(BeNumerically(">", 1))
-
-		// The total size of our cluster is
-		// len(existingMachineSets) + clusterExpansionSize. We
-		// cap that to $max-1 because we want to test that the
-		// maxNodesTotal flag is respected by the
-		// cluster-autoscaler
-		maxNodesTotal := len(nodes) + clusterExpansionSize - 1
-
-		eventWatcher := newEventWatcher(clientset)
-		Expect(eventWatcher.run()).Should(BeTrue())
-		defer eventWatcher.stop()
-
-		// Log cluster-autoscaler events
-		eventWatcher.onEvent(matchAnyEvent, func(e *corev1.Event) {
-			if e.Source.Component == clusterAutoscalerComponent {
-				glog.Infof("%s: %s", e.InvolvedObject.Name, e.Message)
+			// Now that we have created some transient machinesets
+			// take stock of the number of nodes we now have.
+			By("Getting nodes")
+			nodes, err := framework.GetNodes(client)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(nodes)).To(BeNumerically(">=", 1))
+
+			var machineAutoscalers []*caov1beta1.MachineAutoscaler
+
+			By(fmt.Sprintf("Creating %v machineautoscalers", len(machineSets)))
+			transientMachineSetsMinSize := int32(1)
+			var clusterExpansionSize int
```

Comment:
This seems to only be used to check the length of `machineSets` indirectly, is that required? Could we not just assert `machineSets` has a length before doing the range operation?